### PR TITLE
[JENKINS-28454] Pipeline compatibility

### PR DESCRIPTION
**Pull request by alexsomai**
Related to issue [[JENKINS-28454]](https://issues.jenkins-ci.org/browse/JENKINS-28454)

Summary of this pull request:
- [x] Update POM to prepare for Pipeline integration
- [x] Implement `SimpleBuildStep`
- [x] Fix failing test
- [x] Fix FindBugs warnings
- [x] Write Pipeline smoke test
- [x] Use `@DataBoundSetter` to minimize *Snippet Generator* output

### CODE REVIEW

File: src/main/java/hudson/plugins/ws_cleanup/Wipeout.java [hunk 1]
```diff
@@ -79,6 +80,11 @@ public Void invoke(File f, VirtualChannel channel) throws IOException, Interrupt
             }
             return null;
         }
+
+        @Override
+        public void checkRoles(RoleChecker roleChecker) throws SecurityException {
```
Comment (olivergondza) [inline src/main/java/hudson/plugins/ws_cleanup/Wipeout.java]: `MasterToSlaveCallable` should be used here.

File: src/main/java/hudson/plugins/ws_cleanup/Cleanup.java [hunk 2]
```diff
@@ -159,4 +160,9 @@ private void doDelete(List<String> cmdList) throws IOException, InterruptedExcep
     protected void perform(FilePath workspace) throws IOException, InterruptedException {
         workspace.act(this);
     }
+
+    @Override
+    public void checkRoles(RoleChecker roleChecker) throws SecurityException {
```
Comment (olivergondza) [inline src/main/java/hudson/plugins/ws_cleanup/Cleanup.java]: `MasterToSlaveCallable` should be used here.

File: pom.xml [hunk 3]
```diff
@@ -13,6 +13,11 @@
   <version>0.30-SNAPSHOT</version>
   <url>http://wiki.jenkins-ci.org/display/JENKINS/Workspace+Cleanup+Plugin</url>
 
+  <properties>
+    <jenkins.version>1.580.1</jenkins.version>
+    <java.level>7</java.level>
```
Comment (olivergondza) [inline pom.xml]: Is there some reason to drop 6 support?

Comment (alexsomai) [inline pom.xml]: No strong opinion on this one. I can revert back to 6.

**Comment by alexsomai** [comment]
There are some tests failing because of the `matrix-project` dependency, but I don't know how to fix them. Any hints?

**Comment by olivergondza** [comment]
> hudson.util.CyclicGraphDetector$CycleDetectedException: Cycle detected: Plugin:matrix-project -> Plugin:script-security -> Plugin:matrix-auth -> Plugin:windows-slaves -> Plugin:antisamy-markup-formatter -> Plugin:matrix-project

Is caused by incompatible core and matrix plugin version used, I think.

File: src/main/java/hudson/plugins/ws_cleanup/WsCleanup.java [hunk 4]
```diff
@@ -10,59 +10,105 @@
...
-    private boolean cleanWhenUnstable;
-    private boolean cleanWhenFailure;
-    private boolean cleanWhenNotBuilt;
-    private boolean cleanWhenAborted;
+    private boolean skipWhenFailed = false; // keep it for backward compatibility
+    private boolean cleanWhenSuccess = true;
+    private boolean cleanWhenUnstable = true;
+    private boolean cleanWhenFailure = true;
+    private boolean cleanWhenNotBuilt = true;
+    private boolean cleanWhenAborted = true;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/ws_cleanup/WsCleanup.java]: Do we want to keep it enabled by default?

File: src/main/java/hudson/plugins/ws_cleanup/WsCleanup.java [hunk 5]
```diff
@@ -136,6 +187,9 @@ public String getExternalDelete() {
     }
 
     private boolean shouldCleanBuildBasedOnState(Result result) {
+        if (result == null)
+            // in case of Pipeline, the result may be null
+            return true;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/ws_cleanup/WsCleanup.java]: Good practice is to use brackets even for single-line conditions

File: src/main/java/hudson/plugins/ws_cleanup/WsCleanupMatrixAggregator.java [hunk 6]
```diff
@@ -47,6 +48,8 @@ public boolean endBuild() throws InterruptedException, IOException {
     }
 
     private boolean shouldCleanBuildBasedOnState(Result result) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/ws_cleanup/WsCleanupMatrixAggregator.java]: Good practice: Use CheckForNull to document the possible Null parameter for method users

**Comment by oleg-nenashev** [comment]
Generally looks good 👍 
Would be useful to have several tests and documentation drafts for wiki

File: src/main/java/hudson/plugins/ws_cleanup/WsCleanup.java (hunk 4 above)
Comment (alexsomai) [inline src/main/java/hudson/plugins/ws_cleanup/WsCleanup.java]: In the *WsCleanup/config.jelly* they have `default="true"`, so I think it makes sense to match the jelly file.

File: src/main/java/hudson/plugins/ws_cleanup/WsCleanup.java (hunk 5 above)
Comment (alexsomai) [inline src/main/java/hudson/plugins/ws_cleanup/WsCleanup.java]: I do agree for using brackets even for single-line conditions, but in this case I though to be consistent with the existing code. The next if statements don't have brackets. Should I also refactor the next lines within this method?

Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/ws_cleanup/WsCleanup.java]: I commonly try avoiding unrelated changes. But if you change the code

**Comment by alexsomai** [comment]
I have added more tests for Pipeline usage.
Here is a documentation draft that may be used in wiki: https://docs.google.com/document/d/1jeIMxyiUyyv-QKXCg1P5_O0e7j89whIO7WfXLG-kcmk/edit?usp=sharing

File: src/test/java/hudson/plugins/ws_cleanup/CleanupTest.java [hunk 7]
```diff
@@ -36,15 +36,7 @@
...
-import hudson.model.AbstractBuild;
-import hudson.model.BuildListener;
-import hudson.model.Cause;
-import hudson.model.FreeStyleBuild;
-import hudson.model.FreeStyleProject;
-import hudson.model.ParametersAction;
-import hudson.model.ParametersDefinitionProperty;
-import hudson.model.StringParameterDefinition;
-import hudson.model.StringParameterValue;
+import hudson.model.*;
```
Comment (oleg-nenashev) [inline src/test/java/hudson/plugins/ws_cleanup/CleanupTest.java]: It's not the best idea even for the test code, but ok

**Comment by oleg-nenashev** [comment]
re-:+1:. Thanks for writing the tests

**Comment by oleg-nenashev** [comment]
@olivergondza WDYT?

**Comment by olivergondza** [comment]
What would it take to have a nice API for calling the step from pipeline script (`wsCleanup cleanWhenFailure: true`)? I like this otherwise.

**Comment by alexsomai** [comment]
@olivergondza So you are suggesting that it should be a complete new step?
`wsCleanup cleanWhenFailure: true` can be achieved with [JENKINS-29922](https://issues.jenkins-ci.org/browse/JENKINS-29922) and using `@Symbol`.
What about the `patterns`? 
If creating a new step, it means that most of the logic has to be copy/pasted.

**Comment by alexsomai** [comment]
@olivergondza I assume that you've closed this PR by mistake?

**Comment by olivergondza** [comment]
Merging, thanks!
